modules/commerce_bluesnap_currency/src/EventSubscriber/ExchangeRateBluesnap.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+namespace Drupal\commerce_bluesnap_currency\EventSubscriber;
+
+use Drupal\commerce_currency_resolver\ExchangeRateEventSubscriberBase;
+use Drupal\Component\Serialization\Json;
+use Drupal\Core\Config\ConfigFactoryInterface;
+use Drupal\Core\Logger\LoggerChannelFactoryInterface;
+
+/**
+ * Class ExchangeRateBluesnap.
+ *
+ * Fetches all currency conversion rates from bluesnap currency rate API.
+ */
+class ExchangeRateBluesnap extends ExchangeRateEventSubscriberBase {
+
+  /**
+   * The commerce currency resolver config.
+   *
+   * @var \Drupal\Core\Config\ImmutableConfig
+   */
+  protected $config;
+
+  /**
+   * Logger Factory.
+   *
+   * @var \Drupal\Core\Logger\LoggerChannelFactoryInterface
+   */
+  protected $loggerFactory;
+
+  /**
+   * Creates a new ExchangeRateBluesnap object.
+   *
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
+   *   The config factory.
+   * @param \Drupal\Core\Logger\LoggerChannelFactoryInterface $logger_factory
+   *   The logger factory.
+   */
+  public function __construct(
+    ConfigFactoryInterface $config_factory,
+    LoggerChannelFactoryInterface $logger_factory
+  ) {
+    $this->config = $config_factory->get('commerce_currency_resolver.currency_conversion');
+    $this->loggerFactory = $logger_factory->get('commerce_bluesnap_currency');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function apiUrl($mode = NULL) {
+    if ($mode == "sandbox") {
+      return 'https://sandbox.bluesnap.com/services/2/tools/currency-rates';
+    }
+    return 'https://bluesnap.com/services/2/tools/currency-rates';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function sourceId() {
+    return 'exchange_rate_bluesnap';
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getExternalData($base_currency = NULL) {
+    $external_data = [];
+    // Prepare for client.
+    $url = self::apiUrl(
+      $this->config->get('bluesnap')['mode']
+    );
+    $method = 'GET';
+    $options['auth'] = [
+      $this->config->get('bluesnap')['username'],
+      $this->config->get('bluesnap')['password'],
+    ];
+    $options['headers'] = [
+      'Accept' => 'application/json',
+    ];
+
+    if ($this->apiClient($method, $url, $options)) {
+      $response = Json::decode($this->apiClient($method, $url, $options));
+      if (empty($response['currencyRate'])) {
+        $this->loggerFactory->error(
+          'An error occured while fetching exchange rates from bluesnap @url',
+          [
+            '@url' => $url,
+          ]
+        );
+        return $external_data;
+      }
+      // Loop through the api response and build the exchange rate array.
+      foreach ($response['currencyRate'] as $rate) {
+        $code = (string) $rate['quoteCurrency'];
+        $rate = (string) $rate['conversionRate'];
+        $external_data[$code] = $rate;
+      }
+    }
+    return $external_data;
+  }
```

Comment:
**CHANGE REQUEST**

*Performance*

You need to be careful with calling functions multiple times unnecessarily. Store the result in a variable to avoid the overhead. In this case specifically, you're actually making a call to an external API twice which is very inefficient.

*Readability*

Always avoid unnecessary nesting.

```
$raw_response = $this->apiClient($method, $url, $options);
if (!$raw_response) {
  return [];
}

$response = Json::decode($raw_response);
if (empty($response['currencyRate'])) {
  $this->loggerFactory->error(
    'An error occured while fetching exchange rates from bluesnap @url',
    ['@url' => $url]
  );
  return $external_data;
}

// Loop through the api response and build the exchange rate array.
foreach ($response['currencyRate'] as $rate) {
  $code = (string) $rate['quoteCurrency'];
  $rate = (string) $rate['conversionRate'];
  $external_data[$code] = $rate;
}

return $external_data;
```

Reply:
Corrected